truffleruby/src/main/ruby/post-boot/post-boot.rb

```diff
@@ -16,6 +16,27 @@
     require 'rubygems'
   rescue LoadError
   else
+    unless Gem.dir.include?(Truffle::Boot.ruby_home)
+      bad_gem_home = false
+
+      # rbenv does not set GEM_HOME
+      # rbenv-gemset has to be installed which does set GEM_HOME, it's in the subdir of Truffle::Boot.ruby_home
+      # rbenv/versions/<ruby>/gemsets
+      bad_gem_home ||= Gem.dir.include?('rbenv/versions') && !Gem.dir.include?('rbenv/versions/truffleruby')
+
+      # rvm stores gems at .rvm/gems/<ruby>@<gemset-name>
+      bad_gem_home ||= Gem.dir.include?('rvm/gems') && !Gem.dir.include?('rvm/gems/truffleruby')
+
+      # chruby stores gem in ~/.gem/<ruby>/<version>
+      bad_gem_home ||= Gem.dir.include?('.gem') && !Gem.dir.include?('.gems/truffleruby')
```

Comment:
Checking Gem.path will show you the warning is always there,
as `$HOME/.gem/ruby/2.3.0` is always added by RubyGems, unless GEM_PATH is set
And that's an actual problem as it looks through system Ruby gems if there is a system ruby of version 2.3.x.
(For instance, bad extensions, conflicting gem versions, etc)
But probably it is less problematic or likely than a bad value for GEM_HOME, so let's merge for now.
The right fix for this default system Ruby path is changing RUBY_ENGINE.